File: core/src/net/brenig/pixelescape/render/overlay/CountDownOverlay.java
package net.brenig.pixelescape.render.overlay;

import net.brenig.pixelescape.lib.LogHelper;
import net.brenig.pixelescape.lib.Utils;
import net.brenig.pixelescape.screen.GameScreen;

/**
 * Overlay that renders a simple countdown (from 3)<br></br>
 * pauses the game until countdown is finished
 */
public class CountDownOverlay extends Overlay {

	private static final int COUNT_FROM = 3;
	private static final String GO_TEXT = "GO!";
	/**
	 * only used for short countdown
	 */
	private static final String READY_TEXT = "Ready!";

	private boolean isShort;
	private long startedAt = 0;

	public CountDownOverlay(GameScreen screen) {
		super(screen);
		isShort = screen.game.getGameSettings().shortCountdownEnabled();
	}

	@Override
	public void show() {
		super.show();
		startedAt = System.currentTimeMillis();
	}

	@Override
	public void render(float delta) {
		final long timePassed = screen.isScreenPaused() ? 0 : (System.currentTimeMillis() - startedAt);
		final long secondsPassed = (timePassed / 1000L);
		final long secondsRemaining = isShort ? 1 - secondsPassed : COUNT_FROM - secondsPassed;
		if (secondsRemaining <= -1L) {
			//end
			screen.resetToEmptyOverlay();
			return;
		}

		int fractionOfCurrentSecond = ((int) (timePassed % 1000L)) + 1; //fraction of the current second
		if (fractionOfCurrentSecond <= 0) {
			LogHelper.error("Unknown error when calculating passed time!!, / by zero");
			LogHelper.error("timePassed: " + timePassed);
			LogHelper.error("secondsPassed: " + secondsPassed);
			LogHelper.error("secondsRemaining: " + secondsRemaining);
			LogHelper.error("fractionOfCurrentSecond: " + fractionOfCurrentSecond);
			//restore time
			startedAt = System.currentTimeMillis();
			fractionOfCurrentSecond = 1;
		}
		float fontScale = 5F;
		float alpha = 1F;
		if (fractionOfCurrentSecond < 200)
		{
			// big "entry" font scale (15 to 5)
			float anim = Utils.easeOut((float) fractionOfCurrentSecond, 200F);
			fontScale = 15 - anim * 10;
			alpha = 0.75F + anim / 4F;
		}
		else if (fractionOfCurrentSecond > 700)
		{
			// small "vanish" font scale (5 to 2)
			float anim = Utils.easeIn(fractionOfCurrentSecond - 700, 300);
			fontScale = 5 - anim * 3;
			alpha = 1 - anim / 2F;
		}
		if (isShort)
		{
			fontScale *= 0.7F;
		}

		screen.game.getRenderManager().begin();
		screen.game.getFont().setColor(0.2F, 0.8F, 0, alpha);
		screen.game.getFont().getData().setScale(fontScale);

		if (secondsRemaining <= 0) {
			screen.getFontLayout().setText(screen.game.getFont(), GO_TEXT);
		} else if (!isShort) {
			screen.getFontLayout().setText(screen.game.getFont(), "" + secondsRemaining);
		} else {
			screen.getFontLayout().setText(screen.game.getFont(), READY_TEXT);
		}

		final float xPos = screen.world.getWorldWidth() / 2 - screen.getFontLayout().width / 2;
		final float yPos = screen.world.getWorldHeight() / 2 + screen.getFontLayout().height / 2 + screen.getUiPos();

		screen.game.getFont().draw(screen.game.getBatch(), screen.getFontLayout(), xPos, yPos);
	}

	@Override
	public boolean doesPauseGame() {
		return COUNT_FROM - (((int) (System.currentTimeMillis() - startedAt)) / 1000) > 0;
	}

	@Override
	public boolean shouldHideGameUI() {
		return false;
	}

	@Override
	public boolean shouldPauseOnEscape() {
		return true;
	}
}

File: core/src/net/brenig/pixelescape/render/overlay/CountDownOverlay.kt
package net.brenig.pixelescape.render.overlay

import net.brenig.pixelescape.lib.error
import net.brenig.pixelescape.lib.utils.AnimationUtils
import net.brenig.pixelescape.screen.GameScreen

/**
 * Overlay that renders a simple countdown (from 3)<br></br>
 * pauses the game until countdown is finished
 */
class CountDownOverlay(screen: GameScreen) : Overlay(screen) {

    private val isShort: Boolean = screen.game.gameSettings.shortCountdownEnabled
    private var startedAt: Long = 0

    override fun show() {
        super.show()
        startedAt = System.currentTimeMillis()
    }

    override fun render(delta: Float) {
        val timePassed = if (screen.isScreenPaused) 0 else System.currentTimeMillis() - startedAt
        val secondsPassed = timePassed / 1000L
        val secondsRemaining = if (isShort) 1 - secondsPassed else COUNT_FROM - secondsPassed
        if (secondsRemaining <= -1L) {
            //end
            screen.resetToEmptyOverlay()
            return
        }

        var fractionOfCurrentSecond = (timePassed % 1000L).toInt() + 1 //fraction of the current second
        if (fractionOfCurrentSecond <= 0) {
            error("Unknown error when calculating passed time!!, / by zero")
            error("timePassed: " + timePassed)
            error("secondsPassed: " + secondsPassed)
            error("secondsRemaining: " + secondsRemaining)
            error("fractionOfCurrentSecond: " + fractionOfCurrentSecond)
            //restore time
            startedAt = System.currentTimeMillis()
            fractionOfCurrentSecond = 1
        }
        var fontScale = 4f
        var alpha = 1f
        if (fractionOfCurrentSecond < 200) {
            // big "entry" font scale (12 to 3)
            val anim = AnimationUtils.easeOut(fractionOfCurrentSecond.toFloat(), 200f)
            fontScale = 12 - anim * 8
            alpha = 0.75f + anim / 4f
        } else if (fractionOfCurrentSecond > 700) {
            // small "vanish" font scale (3 to 1)
            val anim = AnimationUtils.easeIn((fractionOfCurrentSecond - 700).toFloat(), 300f)
            fontScale = 4 - anim * 3
            alpha = 1 - anim / 2f
        }
        if (isShort) {
            fontScale *= 0.7f
        }

        screen.game.renderManager.begin()
        screen.game.bigFont.setColor(0.2f, 0.8f, 0f, alpha)
        screen.game.bigFont.data.setScale(fontScale)

        if (secondsRemaining <= 0) {
            screen.fontLayout.setText(screen.game.bigFont, GO_TEXT)
        } else if (!isShort) {
            screen.fontLayout.setText(screen.game.bigFont, "" + secondsRemaining)
        } else {
            screen.fontLayout.setText(screen.game.bigFont, READY_TEXT)
        }

        val xPos = screen.world.worldWidth / 2 - screen.fontLayout.width / 2
        val yPos = (screen.world.worldHeight / 2).toFloat() + screen.fontLayout.height / 2 + screen.uiPos.toFloat()

        screen.game.bigFont.draw(screen.game.batch, screen.fontLayout, xPos, yPos)
        screen.game.bigFont.data.setScale(1f)
    }

    override fun doesPauseGame(): Boolean {
        return COUNT_FROM - (System.currentTimeMillis() - startedAt).toInt() / 1000 > 0
    }

    override fun shouldHideGameUI(): Boolean {
        return false
    }

    override fun shouldPauseOnEscape(): Boolean {
        return true
    }

    companion object {

        private const val COUNT_FROM = 3
        private const val GO_TEXT = "GO!"
        /**
         * only used for short countdown
         */
        private const val READY_TEXT = "Ready!"
    }
}
